File: matrq.cpp
#include "matrq.h"

// costruttore a 0 o 1 parametro
matrq::matrq(const int & righe):
    matrice(righe, righe){}

// costruttore
matrq::matrq(const int & righe, std::vector<double> v):
    matrice(righe, righe, v)
{}

// costruttore di copia
matrq::matrq(const matrice & m):
    matrice(m.getC(), m.getC())
{
    for (int i=0; i<m.getC()*m.getR(); ++i){
        getDati()[i]=m.getDati()[i];
    }
}

// funzione che calcola la trasposta di una matrice
matrq* matrq::trasposta(){
    matrq *tmp=new matrq(this->getR());
    for(int i=0; i<this->getR(); ++i){
        for(int j=0; j<this->getR(); ++j){
            (tmp->getDati())[i+j*this->getR()]=(this->getDati())[j+i*this->getR()];
        }
    }
    delete this;
    return tmp;

}

// funzione che calcola il determinante di una matrice
scalare matrq::determinante() const{
    int n=getC();
    double L[n*n], A[n*n], det=1;
    for(int i=0; i<n*n; ++i){
        A[i]=getDati()[i];
    }
    for (int i=0; i<n*n; i++){
        L[i]=0;
    }
    for (int i=0; i<getC(); i++){
        L[i+n*i]=1;
    }
    for(int k=0; k<(n-1); k++){
        for(int i=k+1;i<n; i++){
            if (A[k+n*k]==0){
                throw ecc_math("divisione per zero");
            }
            L[i+n*k]=A[i+n*k]/A[k+n*k];
            for(int j=k; j<n; j++){
                A[i+n*j]=A[i+n*j]-L[i+n*k]*A[k+n*j];
            }
        }
    }
    for(int i=0; i<n; i++){
        det=det*A[i+n*i];
    }
    return det;
}

// funzione che calcola l'a trasposta'inversa di una matrice
matrq* matrq::inversa(){
    int n = getC(), dim=getR()*getC();
    double A[dim], B[dim], p, y;
    std::vector<double> m_tmp;
    for(int i=0;i<dim;++i){
        A[i]=getDati()[i];
    }
    scalare d=determinante();
    if(d.equals(0)){
        throw ecc_math("operazione non possibile");
    }

    for(int i=0;i<n;++i){
        for(int j=0;j<n;++j){
            if(i==j){
                B[i*n+j]=1;
            }
            else{
                B[i*n+j]=0;
            }
        }
    }

    for(int k=0; k<n; ++k){
        p=A[k*n+k];
        for(int j=0;j<n;++j){
            A[k*n+j]=A[k*n+j]/p;
            B[k*n+j]=B[k*n+j]/p;
            }
        for(int i=0;i<n; ++i){
            if(i!=k){
                y=A[i*n+k];
                for(int j=0; j<n;++j){
                    A[i*n+j]=A[i*n+j]-A[k*n+j]*y;
                    if(A[i*n+j]<(1e-009) && A[i*n+j]>(-1e-009)){
                        A[i*n+j]=0;
                    }
                    B[i*n+j]=B[i*n+j]-B[k*n+j]*y;
                    if(B[i*n+j]<(1e-009) && B[i*n+j]>(-1e-009)){
                        B[i*n+j]=0;
                    }
                }
            }
        }
    }
    for(int i=0; i<dim; ++i){
        m_tmp.push_back(B[i]);
    }

    return new matrq(n, m_tmp);
}

File: java/Kalk/matrq.java
package Kalk;

public class matrq extends matrice {

	/** Costruttore di copia*/
	public matrq(matrq m) {
		this(m.getC(),m.getDati());
	}

	/** Costruttore*/
	public matrq(int r, double[] v) {
		super(r, r, v);
	}

	/** Costruttore ad un parametro*/
	public matrq(int r) {
		super(r, r);
	}

	/** Costruttore senza parametri*/
	public matrq() {
		super(0, 0);
	}
	
	/** Funzione che calcola la trasposta*/
	public matrq trasposta() {
	    matrq tmp=new matrq(getR());
	    for(int i=0; i<this.getR(); ++i){
	        for(int j=0; j<this.getR(); ++j){
	            (tmp.getDati())[i+j*this.getR()]=(this.getDati())[j+i*this.getR()];
	        }
	    }
	    return tmp;
	}
	
	/** Funzione che calcola il determinante
	 * throws ErrCalcolo*/
	public scalare determinante() throws ErrCalcolo {
	    int n=getC();
	    double det=1;
	    double[] L, A;
	    L=new double[n*n];
	    A=new double[n*n];
	    for(int i=0; i<n*n; ++i){
	        A[i]=getDati()[i];
	        L[i]=0;
	    }
	    for (int i=0; i<getC(); i++){
	        L[i+n*i]=1;
	    }
	    for(int k=0; k<(n-1); k++){
	        for(int i=k+1;i<n; i++){
	        	if (A[k+n*k]==0) {
    	    		throw new ErrCalcolo("Divisione per zero");
	        	}
	            L[i+n*k]=A[i+n*k]/A[k+n*k];
	            for(int j=k; j<n; j++){
	                A[i+n*j]=A[i+n*j]-L[i+n*k]*A[k+n*j];
	            }
	        }
	    }
	    for(int i=0; i<n; i++){
	        det=det*A[i+n*i];
	    }
		scalare v=new scalare(det);
		return v;
	}
	
	/** Funzione che calcola l'inversa
	 * throws ErrCalcolo*/
	public matrq inversa() throws ErrCalcolo{
		int n = getC(), dim=getR()*getC();
		scalare d=new scalare(0);
		double[] A=new double[dim], B=new double[dim];
		double p, y;
		for(int i=0;i<dim;++i){
				A[i]=getDati()[i];
		}
		try {
			d=determinante();
		}
		catch (ErrCalcolo ec) {
			return this;
		}
		if(d.equals(0)){
			throw new ErrCalcolo("Determinate unguale a zero, calcolo non possibile");
		}
			for(int i=0;i<n;++i){
					for(int j=0;j<n;++j){
							if(i==j){
									B[i*n+j]=1;
							}
							else{
									B[i*n+j]=0;
							}
					}
			}

			for(int k=0; k<n; ++k){
					p=A[k*n+k];
					if(p==0) {
						throw new ErrCalcolo("Divisione per zero");
					}
					for(int j=0;j<n;++j){
							A[k*n+j]=A[k*n+j]/p;
							B[k*n+j]=B[k*n+j]/p;
							}
					for(int i=0;i<n; ++i){
							if(i!=k){
									y=A[i*n+k];
									for(int j=0; j<n;++j){
											A[i*n+j]=A[i*n+j]-A[k*n+j]*y;
											if(A[i*n+j]<(1e-009) && A[i*n+j]>(-1e-009)){
													A[i*n+j]=0;
											}
											B[i*n+j]=B[i*n+j]-B[k*n+j]*y;
											if(B[i*n+j]<(1e-009) && B[i*n+j]>(-1e-009)){
													B[i*n+j]=0;
											}
									}
							}
					}
			}

		return new matrq(n, B);
	}

}
